Panic Using KCP Cluster

KCP provides a cluster example and us-west.yaml and us-east.yaml cluster as soon I apply the cluster.yaml file in examples/cluster.yaml the error of invalid memory address occurs.

Error Log

```
I0313 18:45:10.326714    3578 options.go:145] external host was not specified, using 192.168.0.5
I0313 18:45:10.327141    3578 etcd.go:58] Creating embedded etcd server
{"level":"warn","ts":1647177310.355583,"caller":"fileutil/fileutil.go:57","msg":"check file permission","error":"directory \"/Users/ishansakhuja/Desktop/new/kcp-main/.kcp\" exist, but the permission is \"drwxr-xr-x\". The recommended permission is \"-rwx------\" to prevent possible unprivileged access to the data"}
{"level":"warn","ts":"2022-03-13T18:45:10.609+0530","caller":"etcdserver/metrics.go:224","msg":"failed to get file descriptor usage","error":"cannot get FDUsage on darwin"}
I0313 18:45:11.573985    3578 plugins.go:158] Loaded 1 mutating admission controller(s) successfully in the following order: NamespaceLifecycle.
I0313 18:45:11.574120    3578 plugins.go:158] Loaded 1 mutating admission controller(s) successfully in the following order: NamespaceLifecycle.
W0313 18:45:11.647113    3578 genericapiserver.go:537] Skipping API apiextensions.k8s.io/v1beta1 because it has no resources.
W0313 18:45:12.083452    3578 genericapiserver.go:537] Skipping API authentication.k8s.io/v1beta1 because it has no resources.
W0313 18:45:12.083958    3578 genericapiserver.go:537] Skipping API authorization.k8s.io/v1beta1 because it has no resources.
W0313 18:45:12.084731    3578 genericapiserver.go:537] Skipping API certificates.k8s.io/v1beta1 because it has no resources.
W0313 18:45:12.085375    3578 genericapiserver.go:537] Skipping API coordination.k8s.io/v1beta1 because it has no resources.
W0313 18:45:12.086711    3578 genericapiserver.go:537] Skipping API rbac.authorization.k8s.io/v1beta1 because it has no resources.
W0313 18:45:12.086720    3578 genericapiserver.go:537] Skipping API rbac.authorization.k8s.io/v1alpha1 because it has no resources.
W0313 18:45:12.088818    3578 genericapiserver.go:537] Skipping API flowcontrol.apiserver.k8s.io/v1alpha1 because it has no resources.
W0313 18:45:12.089392    3578 genericapiserver.go:537] Skipping API events.k8s.io/v1beta1 because it has no resources.
I0313 18:45:12.130077    3578 dynamic_serving_content.go:131] "Starting controller" name="serving-cert::/Users/ishansakhuja/Desktop/new/kcp-main/.kcp/apiserver.crt::/Users/ishansakhuja/Desktop/new/kcp-main/.kcp/apiserver.key"
I0313 18:45:12.130092    3578 secure_serving.go:266] Serving securely on [::]:6443
I0313 18:45:12.130122    3578 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I0313 18:45:12.131062    3578 apf_controller.go:317] Starting API Priority and Fairness config controller
I0313 18:45:12.131611    3578 crd_finalizer.go:269] Starting CRDFinalizer
I0313 18:45:12.131680    3578 establishing_controller.go:77] Starting EstablishingController
I0313 18:45:12.131689    3578 nonstructuralschema_controller.go:192] Starting NonStructuralSchemaConditionController
I0313 18:45:12.131694    3578 controller.go:130] Starting OpenAPI controller
I0313 18:45:12.131848    3578 naming_controller.go:296] Starting NamingConditionController
I0313 18:45:12.132047    3578 apiapproval_controller.go:186] Starting KubernetesAPIApprovalPolicyConformantConditionController
I0313 18:45:12.132725    3578 clusterroleaggregation_controller.go:194] Starting ClusterRoleAggregator
I0313 18:45:12.133011    3578 shared_informer.go:240] Waiting for caches to sync for ClusterRoleAggregator
I0313 18:45:12.133454    3578 cluster_authentication_trust_controller.go:440] Starting cluster_authentication_trust_controller controller
I0313 18:45:12.133472    3578 shared_informer.go:240] Waiting for caches to sync for cluster_authentication_trust_controller
I0313 18:45:12.153082    3578 controller.go:147] handling secret "sh.helm.release.v1.fybrik-crd.v1"
I0313 18:45:12.153144    3578 controller.go:147] handling secret "sh.helm.release.v1.fybrik.v1"
I0313 18:45:12.233209    3578 shared_informer.go:247] Caches are synced for ClusterRoleAggregator 
I0313 18:45:12.233588    3578 shared_informer.go:247] Caches are synced for cluster_authentication_trust_controller 
I0313 18:45:13.431953    3578 bootstrap.go:76] bootstrapping clusters.cluster.example.dev
I0313 18:45:13.431964    3578 bootstrap.go:76] bootstrapping negotiatedapiresources.apiresource.kcp.dev
I0313 18:45:13.431951    3578 bootstrap.go:76] bootstrapping workspaceshards.tenancy.kcp.dev
I0313 18:45:13.431959    3578 bootstrap.go:76] bootstrapping apiresourceimports.apiresource.kcp.dev
I0313 18:45:13.431958    3578 bootstrap.go:76] bootstrapping workspaces.tenancy.kcp.dev
I0313 18:45:13.446606    3578 bootstrap.go:78] bootstrapped workspaceshards.tenancy.kcp.dev after 14.786745ms
I0313 18:45:13.446805    3578 bootstrap.go:78] bootstrapped clusters.cluster.example.dev after 15.014912ms
I0313 18:45:13.446870    3578 bootstrap.go:78] bootstrapped workspaces.tenancy.kcp.dev after 15.042838ms
I0313 18:45:13.446880    3578 bootstrap.go:78] bootstrapped apiresourceimports.apiresource.kcp.dev after 15.026178ms
I0313 18:45:13.446923    3578 bootstrap.go:78] bootstrapped negotiatedapiresources.apiresource.kcp.dev after 15.038738ms
I0313 18:45:17.348024    3578 server.go:280] Bootstrapped CRDs and synced all informers. Ready to start controllers
I0313 18:45:17.348062    3578 controller.go:173] Starting WorkspaceShard controller
I0313 18:45:17.348068    3578 shared_informer.go:240] Waiting for caches to sync for workspaceshard
I0313 18:45:17.348073    3578 shared_informer.go:247] Caches are synced for workspaceshard 
I0313 18:45:17.348151    3578 controller.go:192] Starting Workspace controller
I0313 18:45:17.348161    3578 shared_informer.go:240] Waiting for caches to sync for workspace
I0313 18:45:17.348169    3578 shared_informer.go:247] Caches are synced for workspace 
I0313 18:45:17.348185    3578 controller.go:205] Starting Namespace scheduler
I0313 18:45:17.348198    3578 shared_informer.go:240] Waiting for caches to sync for namespace-scheduler
I0313 18:45:17.348206    3578 shared_informer.go:247] Caches are synced for namespace-scheduler 
I0313 18:45:17.348244    3578 namespace_controller.go:200] Starting namespace controller
I0313 18:45:17.348250    3578 shared_informer.go:240] Waiting for caches to sync for namespace
I0313 18:45:17.348258    3578 shared_informer.go:247] Caches are synced for namespace 
I0313 18:45:17.348420    3578 namespace.go:327] Observing Cluster local
I0313 18:45:17.348422    3578 namespace.go:265] Reconciling namespace "admin"|"default"
I0313 18:45:17.348429    3578 namespace.go:265] Reconciling namespace "admin"|"fybrik-blueprints"
I0313 18:45:17.348465    3578 namespace.go:265] Reconciling namespace "admin"|"fybrik-system"
I0313 18:45:17.348467    3578 namespace.go:265] Reconciling namespace "user"|"fybrik-system"
I0313 18:45:17.361636    3578 controller.go:231] Starting Cluster controller
I0313 18:45:17.361662    3578 controller.go:301] Starting APIResource controller
I0313 18:45:17.361696    3578 cluster.go:45] reconciling cluster "local"
E0313 18:45:17.362358    3578 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 6608 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x34bf380, 0x5728a40})
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/runtime/runtime.go:74 +0x7d
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc007e37500})
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/runtime/runtime.go:48 +0x75
panic({0x34bf380, 0x5728a40})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/apis/apiresource/v1alpha1.(*SubResources).ImportFromCRDVersion(0xc002a57a30, 0xc003780810)
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/apis/apiresource/v1alpha1/common_types.go:86 +0x38
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).ensureAPIResourceCompatibility(0xc0014d4d10, {0x3dd7e18, 0xc0037807e0}, {0xc0071d3b72, 0x5}, {{0x0, 0x0}, {0xc0071d3bac, 0x2}, {0xc0071d3bd0, ...}}, ...)
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/negotiation.go:437 +0x1df9
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).process(0xc0014d4d10, {0x3dd7d70, 0xc0044b0880}, {{0x38a5ec7, 0x7}, {0x38bac61, 0x11}, {0xc0012fcba0, 0x1f}, {{0x0, ...}, ...}, ...})
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/negotiation.go:98 +0xb85
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).processNextWorkItem(0xc0014d4d10, {0x3dd7d70, 0xc0044b0880})
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/controller.go:328 +0x1eb
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).startWorker(...)
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/controller.go:312
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).Start.func1()
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/controller.go:305 +0x45
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xd848d80)
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x57afef8, {0x3d84c80, 0xc003780000}, 0x1, 0xc00880f8c0)
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002d079e0, 0x3b9aca00, 0x0, 0xd0, 0xc002d079e0)
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x21e4120, 0xc003649b00, 0xc00292d7b8)
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/wait/wait.go:90 +0x25
created by github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).Start
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/controller.go:305 +0x1a5
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x2b97c58]

goroutine 6608 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc007e37500})
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x34bf380, 0x5728a40})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/apis/apiresource/v1alpha1.(*SubResources).ImportFromCRDVersion(0xc002a57a30, 0xc003780810)
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/apis/apiresource/v1alpha1/common_types.go:86 +0x38
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).ensureAPIResourceCompatibility(0xc0014d4d10, {0x3dd7e18, 0xc0037807e0}, {0xc0071d3b72, 0x5}, {{0x0, 0x0}, {0xc0071d3bac, 0x2}, {0xc0071d3bd0, ...}}, ...)
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/negotiation.go:437 +0x1df9
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).process(0xc0014d4d10, {0x3dd7d70, 0xc0044b0880}, {{0x38a5ec7, 0x7}, {0x38bac61, 0x11}, {0xc0012fcba0, 0x1f}, {{0x0, ...}, ...}, ...})
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/negotiation.go:98 +0xb85
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).processNextWorkItem(0xc0014d4d10, {0x3dd7d70, 0xc0044b0880})
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/controller.go:328 +0x1eb
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).startWorker(...)
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/controller.go:312
github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).Start.func1()
        /Users/ishansakhuja/Desktop/new/kcp-main/pkg/reconciler/apiresource/controller.go:305 +0x45
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xd848d80)
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x57afef8, {0x3d84c80, 0xc003780000}, 0x1, 0xc00880f8c0)
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002d079e0, 0x3b9aca00, 0x0, 0xd0, 0xc002d079e0)
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(0x21e4120, 0xc003649b00, 0xc00292d7b8)
        /Users/ishansakhuja/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220210121228-50b2affb6ca1/pkg/util/wait/wait.go:90 +0x25
created by github.com/kcp-dev/kcp/pkg/reconciler/apiresource.(*Controller).Start
```

I wanted to create a simple Nginx deployment that was given in the Kubernetes official documentation but the kcp is not starting after I have applied the cluster.yaml file

Cluster.yaml components

```apiVersion: cluster.example.dev/v1alpha1
kind: Cluster
metadata:
  name: local
spec:
  kubeconfig: |
```

To reproduce

kubectl apply -f examples/cluster.yaml

--- a/pkg/apis/apiresource/v1alpha1/common_types.go
+++ b/pkg/apis/apiresource/v1alpha1/common_types.go
@@ -83,14 +83,14 @@ func (sr *SubResources) ImportFromCRDVersion(crdVersion *apiextensionsv1.CustomR
 		return false
 	}
 
-	if crdVersion.Subresources.Scale != nil {
+	if crdVersion.Subresources != nil && crdVersion.Subresources.Scale != nil {
 		if !alreadyExists(ScaleSubResourceName) {
 			*sr = append(*sr, SubResource{
 				Name: ScaleSubResourceName,
 			})
 		}
 	}
-	if crdVersion.Subresources.Status != nil {
+	if crdVersion.Subresources != nil && crdVersion.Subresources.Status != nil {
 		if !alreadyExists(StatusSubResourceName) {
 			*sr = append(*sr, SubResource{
 				Name: StatusSubResourceName,

